File: example/rust/src/bin/0x02_read_damiao_pd.rs
use rebotarm_control_rt_rust_examples::common::{
    arg_value, arg_values, has_flag, open_controller, B601_JOINTS,
};
use std::env;
use std::error::Error;
use std::time::Duration;

struct Target {
    label: String,
    port: String,
}

fn usage() {
    println!(
        "Read Damiao POS_VEL gain registers from B601 motors.\n\
Usage:\n\
  cargo run --bin 0x02_read_damiao_pd -- --default-bi [--timeout-ms 300]\n\
  cargo run --bin 0x02_read_damiao_pd -- --port /dev/ttyACM0 [--port /dev/ttyACM1]\n\
\n\
Registers: 25 vel_kp, 26 vel_ki, 27 pos_kp, 28 pos_ki"
    );
}

fn read_target(target: &Target, timeout: Duration) -> Result<(), Box<dyn Error>> {
    println!("\n[{}] {}", target.label, target.port);
    let controller = open_controller(&target.port)?;
    println!(
        "{:<16} {:>10} {:>10} {:>10} {:>10}",
        "joint", "vel_kp", "vel_ki", "pos_kp", "pos_ki"
    );

    for joint in B601_JOINTS {
        let motor = controller.add_motor(joint.motor_id, joint.feedback_id, joint.model)?;
        let values = [25_u8, 26, 27, 28].map(|rid| {
            motor
                .get_register_f32(rid, timeout)
                .map(|v| format!("{v:>10.6}"))
                .unwrap_or_else(|err| format!("{:>10}", format!("ERR({err})")))
        });
        println!(
            "{:<16} {} {} {} {}",
            joint.name, values[0], values[1], values[2], values[3]
        );
    }

    let _ = controller.close_bus();
    Ok(())
}

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().skip(1).collect();
    if has_flag(&args, "--help") || has_flag(&args, "-h") {
        usage();
        return Ok(());
    }

    let timeout_ms = arg_value(&args, "--timeout-ms")
        .and_then(|value| value.parse::<u64>().ok())
        .unwrap_or(300);
    let mut targets = Vec::new();
    if has_flag(&args, "--default-bi") {
        targets.push(Target {
            label: "left".to_string(),
            port: arg_value(&args, "--left-port").unwrap_or_else(|| "/dev/ttyACM0".to_string()),
        });
        targets.push(Target {
            label: "right".to_string(),
            port: arg_value(&args, "--right-port").unwrap_or_else(|| "/dev/ttyACM1".to_string()),
        });
    }
    for (idx, port) in arg_values(&args, "--port").into_iter().enumerate() {
        targets.push(Target {
            label: format!("port{}", idx + 1),
            port,
        });
    }

    if targets.is_empty() {
        usage();
        return Err("no target specified; pass --default-bi or --port".into());
    }

    let timeout = Duration::from_millis(timeout_ms);
    for target in &targets {
        read_target(target, timeout)?;
    }
    Ok(())
}

File: example/cpp/src/0x02_read_damiao_pd.cpp
#include "example_common.hpp"

#include <exception>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

namespace {

struct Target {
  std::string label;
  std::string port;
};

void usage() {
  std::cout
      << "Read Damiao POS_VEL gain registers from B601 motors.\n"
      << "Usage:\n"
      << "  ./0x02_read_damiao_pd --default-bi [--timeout-ms 300]\n"
      << "  ./0x02_read_damiao_pd --port /dev/ttyACM0 [--port /dev/ttyACM1]\n"
      << "\n"
      << "Registers: 25 vel_kp, 26 vel_ki, 27 pos_kp, 28 pos_ki\n";
}

void read_target(const Target& target, uint32_t timeout_ms) {
  std::cout << "\n[" << target.label << "] " << target.port << "\n";
  auto controller = example::open_controller(target.port);
  std::cout << std::left << std::setw(16) << "joint" << std::right << std::setw(12)
            << "vel_kp" << std::setw(12) << "vel_ki" << std::setw(12) << "pos_kp"
            << std::setw(12) << "pos_ki" << "\n";

  for (const auto& joint : example::b601_joints()) {
    auto motor = controller->add_damiao_motor(joint.motor_id, joint.feedback_id, joint.model);
    std::cout << std::left << std::setw(16) << joint.name;
    for (const uint8_t rid : {25, 26, 27, 28}) {
      try {
        std::cout << std::right << std::setw(12) << std::fixed << std::setprecision(6)
                  << motor.get_register_f32(rid, timeout_ms);
      } catch (const std::exception& e) {
        std::string err = "ERR";
        std::cout << std::right << std::setw(12) << err;
      }
    }
    std::cout << "\n";
  }

  try {
    controller->close_bus();
  } catch (...) {
  }
}

}  // namespace

int main(int argc, char** argv) {
  try {
    if (example::has_flag(argc, argv, "--help") || example::has_flag(argc, argv, "-h")) {
      usage();
      return 0;
    }

    const uint32_t timeout_ms =
        static_cast<uint32_t>(example::arg_int(argc, argv, "--timeout-ms", 300));
    std::vector<Target> targets;
    if (example::has_flag(argc, argv, "--default-bi")) {
      targets.push_back({"left", example::arg_value(argc, argv, "--left-port", "/dev/ttyACM0")});
      targets.push_back({"right", example::arg_value(argc, argv, "--right-port", "/dev/ttyACM1")});
    }
    int index = 1;
    for (const auto& port : example::arg_values(argc, argv, "--port")) {
      targets.push_back({"port" + std::to_string(index++), port});
    }
    if (targets.empty()) {
      usage();
      throw std::runtime_error("no target specified; pass --default-bi or --port");
    }
    for (const auto& target : targets) read_target(target, timeout_ms);
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "error: " << e.what() << "\n";
    return 1;
  }
}
